snuba/api/query.py

```diff
@@ -136,32 +146,59 @@ def raw_query(
 
                     except BaseException as ex:
                         error = str(ex)
-                        status = 500
                         logger.exception("Error running query: %s\n%s", sql, error)
+                        stats = log_query_and_update_stats(request, sql, timer, stats, 'error', query_settings)
+                        meta = {}
                         if isinstance(ex, ClickHouseError):
-                            result = {
-                                "error": {
-                                    "type": "clickhouse",
-                                    "code": ex.code,
-                                    "message": error,
-                                }
-                            }
+                            type = "clickhouse"
+                            meta["code"] = ex.code
                         else:
-                            result = {"error": {"type": "unknown", "message": error}}
-
+                            type = "unknown"
+                        raise RawQueryException(
+                            type=type,
+                            message=error,
+                            stats=stats,
+                            timer=timer,
+                            sql=sql,
+                            **meta,
+                        )
             except RateLimitExceeded as ex:
-                error = str(ex)
-                status = 429
-                result = {
-                    "error": {
-                        "type": "ratelimit",
-                        "message": "rate limit exceeded",
-                        "detail": error,
-                    }
-                }
+                stats = log_query_and_update_stats(request, sql, timer, stats, 'rate-limited', query_settings)
+                raise RawQueryException(
+                    type="rate-limited",
+                    message="rate limit exceeded",
+                    stats=stats,
+                    timer=timer,
+                    sql=sql,
+                    detail=str(ex)
+                )
 
-    stats.update(query_settings)
+    stats = log_query_and_update_stats(request, sql, timer, stats, 'success', query_settings)
 
+    result["timing"] = timer
+
+    if settings.STATS_IN_RESPONSE or request.settings.get_debug():
+        result["stats"] = stats
+        result["sql"] = sql
+
+    return result
+
+
+def log_query_and_update_stats(
+    request: Request,
+    sql: str,
+    timer: Timer,
+    stats: MutableMapping,
+    status: str,
+    query_settings: MutableMapping,
```

Comment:
Please provide type hints for the MutableMapping. This for example should be at least `MutableMapping[str, Any]` but likely it can be narrowed down.
Same for the others.